tests/core/kernel-and-base-single-reboot/task.yaml

```diff
@@ -0,0 +1,132 @@
+        echo "This test needs test keys to be trusted"
+        exit
+    fi
+
+    #shellcheck source=tests/lib/store.sh
+    . "$TESTSLIB"/store.sh
+
+    core_snap=core20
+    if os.query is-core18; then
+        core_snap=core18
+    fi
+
+    if [ "$SPREAD_REBOOT" = 0 ]; then
+        init_fake_refreshes "$BLOB_DIR" pc-kernel
+        init_fake_refreshes "$BLOB_DIR" "$core_snap"
+
+        # taken from transition_to_recover_mode()
+        cp /bin/systemctl /tmp/orig-systemctl
+        mount -o bind "$TESTSLIB/mock-shutdown" /bin/systemctl
+        tests.cleanup defer umount /bin/systemctl
+
+        snap refresh --no-wait "$core_snap" pc-kernel > refresh-change-id
+        test -n "$(cat refresh-change-id)"
+        change_id="$(cat refresh-change-id)"
+        # wait until we observe reboots
+        # shellcheck disable=SC2016
+        retry -n 100 --wait 5 sh -c 'test "$(wc -l < /tmp/mock-shutdown.calls)" -gt "1"'
+        # stop snapd now to avoid snapd waiting for too long and deciding to
+        # error out assuming a rollback across reboot
+        systemctl stop snapd.service snapd.socket
+
+        # both link snaps should be done now, snapd was stopped, so we cannot
+        # use 'snap change' and we need to inspect the state directly (even if
+        # snapd was up, it would not respond to API requests as it would be busy
+        # retrying auto-connect)
+        snap debug state --change "$change_id" /var/lib/snapd/state.json > tasks.state
+        # both link snaps are done
+        MATCH ' Done\s+.*Make snap "pc-kernel" .* available' < tasks.state
+        MATCH " Done\s+.*Make snap \"$core_snap\" .* available" < tasks.state
+        # auto-connect of the base is in doing and waiting for reboot
+        MATCH " Doing\s+.*Automatically connect eligible plugs and slots of snap \"$core_snap\"" < tasks.state
+        # auto-connect of the kernel is still queued
+        MATCH ' Do\s+.*Automatically connect eligible plugs and slots of snap "pc-kernel"' < tasks.state
+
+        if os.query is-core18; then
+            snap debug boot-vars > boot-vars.dump
+            MATCH 'snap_mode=try' < boot-vars.dump
+            MATCH 'snap_try_core=core18_.*.snap' < boot-vars.dump
+            MATCH 'snap_try_kernel=pc-kernel_.*.snap' < boot-vars.dump
+        elif os.query is-core20; then
+            stat /boot/grub/kernel.efi | MATCH 'pc_kernel.*.snap/kernel.efi'
+            stat -L /boot/grub/kernel.efi
+            stat /boot/grub/try-kernel.efi | MATCH 'pc_kernel.*.snap/kernel.efi'
+            stat -L /boot/grub/try-kernel.efi
+        else
+            echo "unsupported Ubuntu Core system"
+            exit 1
+        fi
+
+        # restore shutdown so that spread can reboot the host
+        tests.cleanup pop
+
+        REBOOT
+    elif [ "$SPREAD_REBOOT" = 1 ]; then
+        change_id="$(cat refresh-change-id)"
+        # XXX: is this sufficiently robust?
+        snap watch "$change_id" || true
```

Comment:
Will it time-out in a reasonable time if not ready?

Reply:
I think so. I mean, that's what we do in all other tests, so unless it's fundamentally broken we should hit the test timeout eventually.